### ISSUE TITLE
lookahead_not() not matching correctly for non-trivial combinators

### ISSUE THREAD

**Issue by rawtaz** [dashbitco/nimble_parsec#76]
I have the following small code, running with the nimble_parsec dependency at version 0.6:

~~~
defmodule Foo do
  import NimbleParsec

  defparsec :test,
    (
      string("START ")
      |> lookahead_not(string("MIDDLE/"))    # Works
      # |> lookahead_not(concat(ascii_string([?A..?Z], min: 1), string("/")))   # Fails
      # |> lookahead_not(ascii_string([?A..?Z], min: 1) |> string("/"))   # Fails
      |> ascii_string([?A..?Z], [min: 1, max: 7])
    )
end
~~~

This test parsec is supposed to match "START " followed by some A-Z characters that are NOT followed by a "/".

It works for the `lookahead_not()` commented with "# Works" (that is, it tells me "did not expect string ...") but does not work with the `lookahead_not()`s commented with "# Fails" (that is, it matches the A-Z up to but not including the "/", while instead it should tell me "did not expect string ..."):

With `lookahead_not(string("MIDDLE/"))` active:
~~~
4▶ Foo.test("START MIDDLE END") 
{:ok, ["START ", "MIDDLE"], " END", %{}, {1, 0}, 12}
5▶ Foo.test("START MIDDLE/ END")
{:error, "did not expect string \"MIDDLE/\"", "MIDDLE/ END", %{}, {1, 0}, 6}
~~~

With `lookahead_not(concat(ascii_string([?A..?Z], min: 1), string("/")))` active:
~~~
7▶ Foo.test("START MIDDLE END") 
{:ok, ["START ", "MIDDLE"], " END", %{}, {1, 0}, 12}
8▶ Foo.test("START MIDDLE/ END")
{:ok, ["START ", "MIDDLE"], "/ END", %{}, {1, 0}, 12}
~~~

With `lookahead_not(ascii_string([?A..?Z], min: 1) |> string("/"))` active:
~~~
10▶ Foo.test("START MIDDLE END") 
{:ok, ["START ", "MIDDLE"], " END", %{}, {1, 0}, 12}
11▶ Foo.test("START MIDDLE/ END")
{:ok, ["START ", "MIDDLE"], "/ END", %{}, {1, 0}, 12}
~~~

As you can see, in the second and third test, the "/" is seemingly not considered/detected/accounted for by the `lookahead_not()`, or the `lookahead_not()` doesn't seem to have any effect.

Using `ascii_string([?A..?Z]` including with other/more `?x` notations inside it has worked perfectly fine for the rest of the project, which is quite a lot of using it, so it doesn't strike me as a problem with encoding or similar, but more as a bug in the `lookahead_not()` combinator.

Let me know if there's anything I can do to try narrowing this down better!

Edit, in case it's relevant:

~~~
14▶ runtime_info

## System and architecture 

Elixir version:     1.10.2
Erlang/OTP version: 22
ERTS version:       10.7
Compiled for:       x86_64-apple-darwin18.7.0
Schedulers:         4
Schedulers online:  4
~~~

**Comment by josevalim** [comment]
This is behaving as expected. When you say "the next input should not be `ascii_string([?A..?Z], min: 1) |> string("/")`", there is a completely valid answer for this, which is to just parse "M" or just parse "MI" and so on. All of them won't have the trailing "/" at the end and therefore they are all a valid answer for your negative lookahead.

**Comment by rawtaz** [comment]
I understand what you're saying. It's pretty undefined though - it's equally valid to say that since the combinator given to lookahead_not() does indeed match the upcoming text, the lookadhead_not() should trigger. It would perhaps be a good idea to make a note about the "minimal matching" nature of lookahead_not() in the documentation?

This makes it really hard to parse "free text", i.e. where you can't list all the allowed characters because it would be impractical, but instead would list the few non-allowed characters :)

**Comment by josevalim** [comment]
Note you can do `ascii_string(not: ...)`, so there are other ways to do negation. Most times the best way is to just assert what you want, exactly because lookaheads may get expensive. Docs to clarify the current behaviour are definitely welcome though.

**Comment by rawtaz** [comment]
Cheers, I will see about other ways to reach the goal and perhaps file a little PR once I get a more complete picture of the situation. Thanks!
